src/request_reader.rs

```diff
@@ -0,0 +1,163 @@
+use actix_web::http::HeaderMap;
+use actix_web::web::BytesMut;
+use actix_web::{web, HttpRequest};
+use chrono::DateTime;
+use cloudevent::{Event, Payload};
+use futures::{Future, IntoFuture, Stream};
+use std::convert::TryInto;
+use cloudevent::http::*;
+
+macro_rules! unwrap_header {
+    ($headers:expr, $key:expr) => {
+        $headers
+            .get($key)
+            .ok_or(actix_web::error::ErrorBadRequest(format!(
+                "Expecting header {}",
+                $key
+            )))
+            .and_then(|ce| {
+                ce.to_str().map(|s| String::from(s)).map_err(|e| {
+                    actix_web::error::ErrorBadRequest(format!(
+                        "Error while parsing header {}: {}",
+                        $key, e
+                    ))
+                })
+            })
+    };
+}
+
+macro_rules! unwrap_and_remove_header {
+    ($headers:expr, $key:expr) => {{
+        let v = unwrap_header!($headers, $key);
+        $headers.remove($key);
+        v
+    }};
+}
+
+// Possible cases:
+// 1. Content-type exists:
+// 1.1 If application/cloudevents+json -> parse structured
+// 1.2 If other -> parse binary
```

Comment:
Curious about this. Doesn't the Knative implementation of Cloud Events only support binary events over HTTP at the moment?

Reply:
It uses the ce go-sdk that supports all encodings, so technically speaking someone could just trigger the switch to use structured encoding.
Tbh i'm not sure what happens when you send to a channel a message encoded in structured, if knative channels just reencode as binary or it preserves the initial encoding.